feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2021 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package notify
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/labstack/echo/v4"
+
+	"magma/feg/gateway/sbi"
+	n7_client "magma/feg/gateway/sbi/specs/TS29512NpcfSMPolicyControl"
+	"magma/feg/gateway/services/n7_n40_proxy/n7"
+	"magma/feg/gateway/services/session_proxy/relay"
+	"magma/gateway/service_registry"
+	"magma/lte/cloud/go/protos"
+)
+
+const (
+	NOTIFY_PATH      = "/"
+	UnableToDeliver  = 3002
+	EncodedSessionId = "encodedSessionId"
+	ShutdownTimeout  = 10 * time.Second
+)
+
+type NotificationHandler struct {
+	Config        *n7.N7ClientConfig
+	NotifyServer  *echo.Echo
+	cloudRegistry service_registry.GatewayRegistry
+}
+
+func NewNotificationHandler(
+	config *n7.N7ClientConfig,
+	cloudReg service_registry.GatewayRegistry,
+) (*NotificationHandler, error) {
+	handler := &NotificationHandler{
+		Config:        config,
+		NotifyServer:  echo.New(),
+		cloudRegistry: cloudReg,
+	}
+	urlDef, err := url.ParseRequestURI(config.NotifyApiRoot)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing notify api root - %s", err)
+	}
+	basePath := fmt.Sprintf("%s/:%s", urlDef.Path, EncodedSessionId)
+	// Register the handlers
+	handler.NotifyServer.POST(basePath+"/update", handler.PostSmPolicyUpdateNotification)
+	handler.NotifyServer.POST(basePath+"/terminate", handler.PostSmPolicyTerminateNotification)
+
+	return handler, nil
+}
+
+func (handler *NotificationHandler) Start() (string, error) {
+	errChan := make(chan error)
+	addr, err := net.ResolveTCPAddr("tcp", handler.Config.LocalAddr)
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		err = handler.NotifyServer.Start(addr.String())
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	return sbi.WaitForSbiServerAndGetAddr(handler.NotifyServer, errChan)
+}
+
+func (handler *NotificationHandler) Shutdown() {
+	ctx, cancel := context.WithTimeout(context.Background(), ShutdownTimeout)
+	defer cancel()
+	handler.NotifyServer.Shutdown(ctx)
+}
```

Comment:
It makes sense but , as mentioned, try to encapsulate it

Reply:
now I understand start and and stop.

I think this would look nicer if we create a sbi client. This wway we will be able to reuse it on other interfaces.
Functions to include there would be start, stop,  WaitForSbiServerAndGetAddr

We did that for [GTP](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Amagma%2Fmagma.git&branch=master&file=feg%2Fgateway%2Fgtp%2Fgtp_client.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.2.2&start_row=51&start_col=11&end_row=51&end_col=11) client